### ISSUE TITLE
Can an objects have attributes as well as properties?

### ISSUE THREAD

**Issue by hadley** [RConsortium/OOP-WG#49]
i.e. can a user still add arbitrary metdata to an object with `attr<-`?

**Comment by mmaechler** [comment]
Good Q:  As you know  `attr<-`  is a primitive *and* its a function that really must remain very fast and may not "support" an internal check for an `R7` object ...  or it may, if we get @ltierney to agree ;-)  and set things up so the check can happen very quickly ... ?

**Comment by mmaechler** [comment]
... or maybe  `properties` should really be attributes internally anyway, and we rely on validity checking... 
rebellious users can muck up the objects in any case, using C, or even by using one of those "low-access" packages (using C) which some people like to put in the hands of the users ...

**Comment by lawremi** [comment]
Yes, most properties would be implemented as attributes, but in principle they could be implemented any number of ways. For example, a property could be dynamically computed from two attributes, or from part of an attribute, or retrieved from a database. 

This comes down to how we define validity. A validity function could operate at the property level, or it could check the validity of the implementation, looking at attributes or whatever. Property-based validation would be robust to implementation changes, while low-level validation would catch errors in the implementation earlier, potentially catching extraneous attributes.

S4 just checks the known slot types and ignores any additional attributes, which is analogous to property-based validation. Extra attributes routinely arise when extending an "old" (S3) class with an S4 class. For example, a class extending a vector class like "character" would construct instances that could have a "names" attribute. While `setOldClass()` does have an `S4Class=` argument for defining slots for attributes known to always be on the parent S3 object, that obviously does not apply to "names". 

With R7 properties, one could define a "names" property of type `Nullable("character")` with an accessor that just delegates to `names()`, so maybe R7 could be stricter than S4, but I would err on the side of flexibility, especially since we're trying to ease the transition from S3 to R7.
